### PR TITLE
ci: 'downgrade' Ubuntu version to Bionic

### PR DESCRIPTION
**Pull request by mrc0mmand**
Since the current Ubuntu Focal image in GH Actions is missing the
`kernel-modules-extra` package with the `scsi_debug` module[0], let's
change the distro flag from `ubuntu-20.04` (Focal) to `ubuntu-latest`
(currently Bionic, as Focal is still in the beta phase).

[0] https://github.community/t/missing-scsi-debug-linux-kernel-module-on-gh-actions/157698

### CODE REVIEW

**Comment by mrc0mmand** [comment]
The "skip ASan's link order verification" commit extends the set of executed tests, since without it almost all tests which use the `scsi_debug` module are skipped:

```
==14386==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
eject: unmount of `/home/runner/work/util-linux/util-linux/tests/output/eject/umount-by-disk-mounted-mnt' failed


diff-{{{
--- /home/runner/work/util-linux/util-linux/tests/expected/eject/umount-by-disk-mounted	2021-01-29 19:00:21.916066535 +0000
+++ /home/runner/work/util-linux/util-linux/tests/output/eject/umount-by-disk-mounted	2021-01-29 19:07:12.435761182 +0000
@@ -1 +1,2 @@
-Success
+rmmod failed: 'modprobe: FATAL: Module scsi_debug is in use.'
+timeout removing scsi_debug module (rmmod)
}}}-diff

 FAILED (eject/umount-by-disk-mounted)
                : by-disk-mounted-partition   ...[14169 umount] scsi_debug already locked!
 SKIPPED (cannot remove scsi_debug module (rmmod))
        fdisk: align 512/4K                   ... SKIPPED (cannot remove scsi_debug module (rmmod))
        fdisk: align 512/4K +alignment_offset ... SKIPPED (cannot remove scsi_debug module (rmmod))
        fdisk: align 512/4K +MD               ... SKIPPED (cannot remove scsi_debug module (rmmod))
...

      libmount: context                        ... SKIPPED (cannot remove scsi_debug module (rmmod))
     libmount: context-py                     ... SKIPPED (pylibmount not compiled)
     libmount: context (utab)                 ... SKIPPED (cannot remove scsi_debug module (rmmod))
     libmount: context-py (utab)              ... SKIPPED (pylibmount not compiled)
...
    etc.
```

**Comment by mrc0mmand** [comment]
After I switched to Ubuntu Bionic I noticed that the `scsi_debug` related tests are still being skipped under sanitizers, so I delved a bit deeper. 

Apart from several straightforward fix-ups, the *ci: deal with uninstrumented binaries using instrumented libs* commit hides a piece of black magic, which deals with the runtime library requirements for ASan. As the commit states, the `eject` tests break ASan, since they `execl()`s `/bin/umount` which is uninstrumented, but it picks up the instrumented `libblkid`. In case of gcc the workaround is quite easy, however clang requires a bit more work.

I used some pieces from systemd CI, since we had to deal with this issue there as well, and some macros from systemd (introduced by @evverx), to correctly detect ASan when compiled with clang.

I tried to keep the ASan-related shenanigans as small as possible, but if you (@karelzak) or @evverx notice any way of improving this, I'm all ears!

File: include/env.h [hunk 1]
```diff
@@ -4,6 +4,23 @@
...
+#  elif defined(__has_feature)
+#    if __has_feature(address_sanitizer)
+#      define HAS_FEATURE_ADDRESS_SANITIZER 1
+#    endif
+#  endif
+#  if !defined(HAS_FEATURE_ADDRESS_SANITIZER)
+#    define HAS_FEATURE_ADDRESS_SANITIZER 0
+#  endif
+#endif
+
```
Comment (karelzak) [inline include/env.h:23]: It would be better to keep this in include/c.h where we have random fallbacks.

File: .github/workflows/cibuild.sh [hunk 2]
```diff
@@ -5,11 +5,11 @@ COMPILER="${COMPILER:?}"
...
-    CC="clang"
-    CXX="clang++"
+    CC="clang-$COMPILER_VERSION"
+    CXX="clang++-$COMPILER_VERSION"
 elif [[ "$COMPILER" == gcc ]]; then
-    CC="gcc"
-    CXX="g++"
+    CC="gcc-$COMPILER_VERSION"
+    CXX="g++-$COMPILER_VERSION"
 fi
```
Comment (karelzak) [inline .github/workflows/cibuild.sh:15]: This (gcc-<version> is Ubuntu specific.

Please, use $COMPILER_VERSION only if defined. The goal is to keep cibuild.sh portable -- maybe one day we will use it also for other distros.

File: .github/workflows/cibuild.yml [hunk 3]
```diff
@@ -17,8 +17,10 @@ jobs:
       fail-fast: false
       matrix:
         env:
-          - { COMPILER: "gcc",   COMPILER_VERSION: "10" }
-          - { COMPILER: "clang", COMPILER_VERSION: "10" }
+          - { COMPILER: "gcc",   COMPILER_VERSION: "10", SANITIZE: "yes"  }
+          - { COMPILER: "gcc",   COMPILER_VERSION: "10", SANITIZE: "no"   }
```
Comment (karelzak) [inline .github/workflows/cibuild.yml:25]: Good idea to add SANITIZE to the matrix. Thanks.

File: include/env.h (hunk 1 above)
Comment (mrc0mmand) [reply include/env.h]: That's indeed way better, thanks!

File: .github/workflows/cibuild.sh (hunk 2 above)
Comment (mrc0mmand) [reply .github/workflows/cibuild.sh:15]: Understood, fixed.

**Comment by karelzak** [comment]
Thanks!

**Pull request merged**